SCons/Tool/MSCommon/MSVC/Dispatcher.py

```diff
@@ -0,0 +1,63 @@
+# MIT License
+#
+# Copyright The SCons Foundation
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
+# KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+"""
+Internal method dispatcher for Microsoft Visual C/C++.
+"""
+
+import sys
+
+from ..common import (
+    debug,
+)
+
+
+_refs = []
+
+def register_class(ref):
+    _refs.append(ref)
+
```

Comment:
what is all this logic for?
What does it do?

Reply:
The modules/classes can be registered.  Any registered modules/classes with reset methods are invoked when MSVC.reset() is called from  reset_installed_vcs() in vc.py.  "Reset" the MSVC reset dispatches calls to those modules/classes registered than have callable reset methods.  This way, child modules can be indirectly invoked without having to hard-code the calls in vc.py.

Modules tend to registered immediately after the Dispatcher import near the top of the file for the files under MSVC.

Similarly, at the end of MSVC \_\_init\_\_.py the dispatcher validate method is invoked which calls all registered module/class validate methods which is used as an internal sanity check that internal data structures have expected keys etc since the data definitions are no longer in a single file.  Usually to verify that a dictionary supports all of the expected keys.  In this way, when a new item (e.g., version) is added, if it is not configures in all associated modules an exception should be raised.

Reply:
register_module/register_class called from modules under MSVC.  reset() called from __init__.py reset method.  validate() called from end of __init__.py initialization.  MSVC.reset() is the only reset/validate call outside/above MSVC.

Reply:
Can you add that to the docstring for this module?
If it wasn't obvious to me now, it's likely to be less obvious in 2023 when the next MSVC comes out.. ;)

Reply:
Done.

Note; the verify invocation was moved from the end of MSVC.\_\_init\_\_.py to the end of vc.py.  A verify method was added to Config to check that all msvc versions defined in vc._VCVER are configured in Config.  This would catch a new version added in vc.py but not in the MSVC modules.